homeassistant/components/omnilogic/sensor.py

```diff
@@ -0,0 +1,688 @@
+        sensordata: dict,
+    ):
+        """Initialize the sensor."""
+        super().__init__(
+            coordinator=coordinator,
+            kind=kind,
+            name=name,
+            backyard=backyard,
+            bow=bow,
+            device_class=device_class,
+            icon=icon,
+            unit=unit,
+            sensordata=sensordata,
+        )
+
+    @property
+    def state(self):
+        """Return the state for the pump speed sensor."""
+        _LOGGER.debug("Updating state of sensor: %s", self._name)
+
+        sensordata = {}
+        pump_type = ""
+        pump_speed = 0
+
+        if self._kind == "filter_pump_speed":
+            for backyard in self.coordinator.data:
+                for bow in backyard.get("BOWS"):
+                    if bow["Filter"]["systemId"] == self.entitydata.get("systemId"):
+                        sensordata = bow["Filter"]
+                        break
+
+            if sensordata.get("Filter-Type") == "FMT_VARIABLE_SPEED_PUMP":
+                pump_type = "VARIABLE"
+            elif sensordata.get("Filter-Type") == "FMP_SINGLE_SPEED":
+                pump_type = "SINGLE"
+            elif sensordata.get("Filter-Type") == "FMT_DUAL_SPEED":
+                pump_type = "DUAL"
+
+            pump_speed = sensordata.get("filterSpeed")
+
+        elif self._kind == "pump_speed":
+            for backyard in self.coordinator.data:
+                for bow in backyard.get("BOWS"):
+                    for pump in bow.get("Pumps"):
+                        if pump["systemId"] == self.entitydata["systemId"]:
+                            sensordata = pump
+                            break
+
+            if sensordata.get("Filter-Type") == "PMP_VARIABLE_SPEED_PUMP":
+                pump_type = "VARIABLE"
+            elif sensordata.get("Filter-Type") == "PMP_SINGLE_SPEED":
+                pump_type = "SINGLE"
+            elif sensordata.get("Filter-Type") == "PMP_DUAL_SPEED":
+                pump_type = "DUAL"
+
+            pump_speed = sensordata.get("pumpSpeed")
+
+        if pump_type == "VARIABLE":
+            self._unit = "%"
+            self._state = pump_speed
+        elif pump_type == "DUAL":
+            if pump_speed == 0:
+                self._state = "off"
+            elif pump_speed == sensordata.get("Min-Pump-Speed"):
+                self._state = "low"
+            elif pump_speed == sensordata.get("Max-Pump-Speed"):
+                self._state = "high"
+        elif pump_type == "SINGLE":
+            if pump_speed == 0:
+                self.state = "off"
+            elif pump_speed == sensordata.get("Max-Pump-Speed"):
+                self._state = "on"
+
+        self._name = (
+            self._backyard["BackyardName"]
+            + " "
+            + self.bow.get("Name")
+            + " "
+            + self.entitydata.get("Name")
+        )
+
+        self._attrs["PumpType"] = pump_type
+
+        super().add_alarms(sensordata.get("Alarms"))
```

Comment:
We don't allow side effects in state properties. I'd probably define an entity callback method that is called when there's new data from the coordinator. Do all the data parsing and attribute updating in that callback. In the `state` property just return `self._state`.

Reply:
Not sure of your suggestion here. Can you clarify?

Reply:
- Add a new entity method that does the data lookup and any side effects. At the end in the method call `self.async_write_ha_state`.
- Register this new method as a callback for the coordinator with `self.coordinator.async_add_listener`. Do this within entity coroutine method `async_added_to_hass`.

See eg:
https://github.com/home-assistant/core/blob/663245c3515d0b44fac08c5e6c8ea29675e2b6ba/homeassistant/helpers/update_coordinator.py#L212-L217

Reply:
OK - since we will be adding a binary sensor later for the alarms on the system, I've removed the alarms as variable data on each element. This was a nice-to-have and we can look at in a future version.